File: java/src/main/java/deltix/luminary/EnumerationDef.java
package deltix.luminary;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public class EnumerationDef extends ItemDef implements TypeDef<TypeEnumeration> {
    private final IntegralType underlyingType;
    private final FileDef file;
    private final TypeScope parent;
    private final List<EnumerationMemberDef> members = new ArrayList<>();
    private final TypeEnumeration type;

    public EnumerationDef(@NotNull TypeScope parent, @NotNull String name, @NotNull IntegralType underlyingType,
                          @Nullable List<String> comments) {
        super(name, parent instanceof FileDef ? name : ((ItemDef) parent).getFullName() + "." + name, comments);
        this.underlyingType = underlyingType;
        this.parent = parent;
        this.file = parent.getFile();
        this.type = new TypeEnumeration(this);

        parent.getDefinedEnumerations().add(this);
    }

    public EnumerationDef(@NotNull TypeScope parent, @NotNull String name, @NotNull IntegralType underlyingType) {
        this(parent, name, underlyingType, null);
    }

    @NotNull
    public IntegralType getUnderlyingType() {
        return underlyingType;
    }

    @Override
    @NotNull
    public FileDef getFile() {
        return file;
    }

    @Override
    @NotNull
    public TypeScope getParent() {
        return parent;
    }

    @NotNull
    public List<EnumerationMemberDef> getMembers() {
        return members;
    }

    @Override
    @NotNull
    public TypeEnumeration getType() {
        return type;
    }
}

File: csharp/main/EnumerationDef.cs
using System;
using System.Collections.Generic;
using JetBrains.Annotations;

namespace Deltix.Luminary
{
	public sealed class EnumerationDef : ItemDef, ITypeDef<TypeEnumeration>
	{
		public EnumerationDef([NotNull] ITypeScope parent, [NotNull] String name, IntegralType underlyingType, [CanBeNull, ItemNotNull] List<String> comments = null)
			: base(name, comments)
		{
			UnderlyingType = underlyingType;
			Parent = parent;

			File = parent.File;
			Members = new List<EnumerationMemberDef>();
			Type = new TypeEnumeration(this);
			FullName = parent is FileDef ? Name : $"{((ItemDef) parent).FullName}.{Name}";

			parent.DefinedEnumerations.Add(this);
		}

		public IntegralType UnderlyingType { get; }

		[NotNull]
		public FileDef File { get; }

		[NotNull]
		public ITypeScope Parent { get; }

		[NotNull, ItemNotNull]
		public List<EnumerationMemberDef> Members { get; }

		[NotNull]
		public TypeEnumeration Type { get; }

		public override String FullName { get; }
	}
}
